kai/jsonrpc/streams.py

```diff
@@ -95,155 +138,179 @@ class LspStyleStream(JsonRpcStream):
 
-            try:
-                if "method" in msg_dict:
-                    return JsonRpcRequest.model_validate(msg_dict)
-                else:
-                    return JsonRpcResponse.model_validate(msg_dict)
-            except Exception as e:
-                return JsonRpcError(
-                    code=JsonRpcErrorCode.ParseError,
-                    message=f"Could not validate JSON: {e}",
-                )
+        try:
+            if "method" in msg_dict:
+                return JsonRpcRequest.model_validate(msg_dict)
+            else:
+                return JsonRpcResponse.model_validate(msg_dict)
+        except Exception as e:
+            return JsonRpcError(
+                code=JsonRpcErrorCode.ParseError,
+                message=f"Could not validate JSON: {e}",
+            )
 
 
 class BareJsonStream(JsonRpcStream):
-    def __init__(
-        self,
-        recv_file: BufferedReader,
-        send_file: BufferedWriter,
-        json_dumps_kwargs: dict[Any, Any] | None = None,
-        json_loads_kwargs: dict[Any, Any] | None = None,
-        log: KaiLogger | None = None,
-    ):
-        super().__init__(recv_file, send_file, json_dumps_kwargs, json_loads_kwargs)
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
 
         self.buffer: str = ""
         self.decoder = json.JSONDecoder()
         self.chunk_size = 512
-        self.log = get_logger("jsonrpc")
-        if log is not None:
-            self.log = log
 
-    def send(self, msg: JsonRpcRequest | JsonRpcResponse) -> None:
+    async def send(self, msg: JsonRpcRequest | JsonRpcResponse) -> None:
+        writer = await self.get_writer()
+
         json_req = f"{msg.model_dump_json(exclude_none=True)}\n"
 
-        log.log(TRACE, "Sending request: %s", dump_json_no_infinite_recursion(msg))
+        if not writer.is_closing():
+            log.log(TRACE, "Sending request: %s", dump_json_no_infinite_recursion(msg))
 
-        with self.send_lock:
-            self.send_file.write(json_req.encode())
-            self.send_file.flush()
+            writer.write(json_req.encode("utf-8"))
+            # TODO(JonahSussman): Do we need this? If you drain a closed stream,
+            # you get an exception.
+            # await writer.drain()
+        else:
+            log.error("Writer is closed or closing, cannot send message")
 
-    def readline(self) -> bytes | None:
-        try:
-            return self.recv_file.readline()
-        except CancelledError:
-            self.log.info("cancelling readline")
-            raise
+    async def unlimited_readline(self, reader: asyncio.StreamReader) -> bytes:
+        result = b""
 
-    def recv(self) -> JsonRpcError | JsonRpcRequest | JsonRpcResponse | None:
-        with self.recv_lock:
-            if self.recv_file.closed:
+        while True:
+            try:
+                result += await reader.readuntil(b"\n")
```

Comment:
wouldnt this reader have the same problem as above?

Reply:
No because once a stream is exhausted, the StreamReader will reliably return EOF https://docs.python.org/3/library/asyncio-stream.html#asyncio.StreamReader